File: remote/src/main/java/com/kelsos/mbrc/ui/activities/ConnectionManagerActivity.java
package com.kelsos.mbrc.ui.activities;

import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.afollestad.materialdialogs.MaterialDialog;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.adapters.ConnectionSettingsAdapter;
import com.kelsos.mbrc.constants.UserInputEventType;
import com.kelsos.mbrc.domain.ConnectionSettings;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.events.ui.ConnectionSettingsChanged;
import com.kelsos.mbrc.events.ui.DiscoveryStopped;
import com.kelsos.mbrc.events.ui.NotifyUser;
import com.kelsos.mbrc.ui.dialogs.SettingsDialogFragment;
import com.kelsos.mbrc.utilities.RxBus;

public class ConnectionManagerActivity extends RoboAppCompatActivity
    implements SettingsDialogFragment.SettingsDialogListener {

  @Inject private RxBus bus;
  @Bind(R.id.connection_list) RecyclerView mRecyclerView;
  @Bind(R.id.toolbar) Toolbar mToolbar;

  private MaterialDialog mProgress;
  private Context mContext;

  @OnClick(R.id.connection_add) public void onAddButtonClick(View v) {
    SettingsDialogFragment settingsDialog = new SettingsDialogFragment();
    Bundle args = new Bundle();
    args.putInt("index", -1);
    settingsDialog.setArguments(args);
    settingsDialog.show(getSupportFragmentManager(), "settings_dialog");
  }

  @OnClick(R.id.connection_scan) public void onScanButtonClick(View v) {
    MaterialDialog.Builder mBuilder = new MaterialDialog.Builder(mContext);
    mBuilder.title(R.string.progress_scanning);
    mBuilder.content(R.string.progress_scanning_message);
    mBuilder.progress(true, 0);
    mProgress = mBuilder.show();
    bus.post(MessageEvent.newInstance(UserInputEventType.StartDiscovery));
  }

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.ui_activity_connection_manager);
    ButterKnife.bind(this);
    setSupportActionBar(mToolbar);
    mRecyclerView.setHasFixedSize(true);
    RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(this);
    mRecyclerView.setLayoutManager(mLayoutManager);
  }

  @Override protected void onStart() {
    super.onStart();
    mContext = this;
    ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true);
      actionBar.setTitle(R.string.connection_manager_title);
    }

  }

  @Override protected void onResume() {
    super.onResume();
    bus.register(ConnectionSettingsChanged.class, this::handleConnectionSettingsChange, false);
    bus.register(DiscoveryStopped.class, this::handleDiscoveryStopped, false);
    bus.register(NotifyUser.class, this::handleUserNotification, false);
  }

  @Override protected void onPause() {
    super.onPause();
    bus.unregister(this);
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        onBackPressed();
        break;
      default:
        return false;
    }
    return true;
  }

  @Override
  public void onDialogPositiveClick(SettingsDialogFragment dialog, ConnectionSettings settings) {
    bus.post(settings);
  }

  public void handleConnectionSettingsChange(ConnectionSettingsChanged event) {
    ConnectionSettingsAdapter mAdapter = new ConnectionSettingsAdapter(event.getSettings(), bus);
    mAdapter.setDefaultIndex(event.getDefaultIndex());
    mRecyclerView.setAdapter(mAdapter);
  }

  public void handleDiscoveryStopped(DiscoveryStopped event) {

    if (mProgress != null) {
      mProgress.dismiss();
    }

    String message;
    switch (event.getReason()) {
      case NO_WIFI:
        message = getString(R.string.con_man_no_wifi);
        break;
      case NOT_FOUND:
        message = getString(R.string.con_man_not_found);
        break;
      case COMPLETE:
        message = getString(R.string.con_man_success);
        break;
      default:
        message = getString(R.string.unknown_reason);
        break;
    }

    Snackbar.make(mRecyclerView, message, Snackbar.LENGTH_SHORT).show();
  }

  public void handleUserNotification(NotifyUser event) {
    final String message =
        event.isFromResource() ? getString(event.getResId()) : event.getMessage();

    Snackbar.make(mRecyclerView, message, Snackbar.LENGTH_SHORT).show();
  }
}

File: app/src/main/java/com/kelsos/mbrc/features/settings/ConnectionManagerActivity.kt
package com.kelsos.mbrc.features.settings

import android.os.Bundle
import android.widget.Button
import androidx.core.view.isGone
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.google.android.material.snackbar.Snackbar
import com.kelsos.mbrc.CommonToolbarActivity
import com.kelsos.mbrc.R
import com.kelsos.mbrc.networking.discovery.DiscoveryStop
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel

class ConnectionManagerActivity :
  CommonToolbarActivity(R.layout.activity_connection_manager),
  SettingsDialogFragment.SettingsSaveListener,
  ConnectionChangeListener {
  private val viewModel: ConnectionManagerViewModel by viewModel()

  private lateinit var recyclerView: RecyclerView
  private lateinit var scanButton: Button

  private val adapter: ConnectionAdapter = ConnectionAdapter()

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    recyclerView = findViewById(R.id.connection_list)
    scanButton = findViewById(R.id.connection_scan)
    findViewById<Button>(R.id.connection_add).setOnClickListener {
      val settingsDialog = SettingsDialogFragment()
      settingsDialog.show(this.supportFragmentManager, "settings_dialog")
    }
    scanButton.setOnClickListener {
      findViewById<LinearProgressIndicator>(R.id.connection_manager__progress).isGone = false
      scanButton.isEnabled = false
      viewModel.actions.startDiscovery()
    }

    recyclerView.setHasFixedSize(true)
    val layoutManager = LinearLayoutManager(this)
    recyclerView.layoutManager = layoutManager
    adapter.setChangeListener(this)
    recyclerView.adapter = adapter
    setToolbarTitle(R.string.connection_manager_title)

    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        viewModel.state.settings.collect {
          adapter.submitData(it)
        }
      }
    }

    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        viewModel.state.events.collect {
          onDiscoveryStopped(it)
        }
      }
    }
  }

  override fun onSave(settings: ConnectionSettings) {
    viewModel.actions.save(settings)
  }

  private fun onDiscoveryStopped(event: DiscoveryStop) {
    findViewById<LinearProgressIndicator>(R.id.connection_manager__progress).isGone = true
    scanButton.isEnabled = true

    val message: String =
      when (event) {
        DiscoveryStop.NoWifi -> getString(R.string.con_man_no_wifi)
        DiscoveryStop.NotFound -> getString(R.string.con_man_not_found)
        is DiscoveryStop.Complete -> {
          getString(R.string.con_man_success)
        }
      }

    Snackbar.make(recyclerView, message, Snackbar.LENGTH_SHORT).show()
  }

  override fun onDelete(settings: ConnectionSettings) {
    viewModel.actions.delete(settings)
  }

  override fun onEdit(settings: ConnectionSettings) {
    val settingsDialog = SettingsDialogFragment.newInstance(settings)
    val fragmentManager = supportFragmentManager
    settingsDialog.show(fragmentManager, "settings_dialog")
  }

  override fun onDefault(settings: ConnectionSettings) {
    viewModel.actions.setDefault(settings)
  }
}
